experimenter/experimenter/nimbus_ui/templatetags/nimbus_extras.py

```diff
@@ -97,3 +98,61 @@ def can_review_experiment(context, experiment):
 @register.filter
 def qa_icon_info(value):
     return NimbusConstants.QAStatus.get_icon_info(value)
+
+
+@register.filter
+def channel_icon_info(value):
+    return NimbusExperiment.Channel.get_icon_info(value)
+
+
+@register.simple_tag
+def render_channel_icons(experiment):
+    channel_items = []
+
+    if experiment.is_desktop and experiment.channels:
+        # For multi-channel desktop experiments, show icon + name for each channel
+        for channel in sorted(experiment.channels):
+            icon_info = NimbusExperiment.Channel.get_icon_info(channel)
+            channel_label = experiment.Channel(channel).label
+            channel_items.append(
+                f'<span class="d-inline-flex align-items-center me-2 mb-1">'
+                f'<span class="{icon_info["color"]} me-1">'
+                f'<i class="{icon_info["icon"]}"></i></span>'
+                f'<span class="small">{channel_label}</span>'
+                f"</span>"
+            )
+    elif experiment.channel:
+        # For single channel experiments
+        icon_info = NimbusExperiment.Channel.get_icon_info(experiment.channel)
+        channel_label = experiment.Channel(experiment.channel).label
+        channel_items.append(
+            f'<span class="d-inline-flex align-items-center">'
+            f'<span class="{icon_info["color"]} me-1">'
+            f'<i class="{icon_info["icon"]}"></i></span>'
+            f'<span class="small">{channel_label}</span>'
+            f"</span>"
+        )
```

Comment:
It's not really a great pattern to construct html within python space like this.  Then you start to get half your templates in template space and half in python space and it starts to get hard to reason about.  Another pattern is to use the template tags to build nested datastructures like lists and dicts that are easy to walk through in template space and construct the HTML there.

Reply:
you can put the HTML bits in a separate template and have the templatetag render that template with specific kwargs